// omni-app-container/demo/my-app.js
import { html, OmniElement, OmniStyleElement } from 'omni-ui';
import { OmniAppContainerMixin } from 'omni-app-container';
import './overview-screen.js';
import './view-context.js';
import './view-helpers.js';
import './view-an-sso.js';

OmniStyleElement.register();

export class MyApp extends OmniAppContainerMixin(OmniElement) {
  static properties = {
    tab: this.routeParamProperty({ name: 'tab' }),
  };

  static routes = baseRoute => [
    {
      path: '/',
      redirect: `${baseRoute}/overview`,
    },
    {
      /* This syntax defines a parameter "tab" which matches a specific pattern "overview". Without the (pattern) :tab would match any string. */
      path: '/:tab(overview)',
      component: 'overview-screen',
    },
    {
      path: '/:tab(context)',
      component: 'view-context',
    },
    {
      path: '/:tab(helpers)',
      component: 'view-helpers',
    },
    {
      path: '/:tab(an-sso)',
      component: 'view-an-sso',
    },
  ];

  render() {
    return html`
      <omni-style>
        <button
          class="button mt-5 ml-5 ${this.tab === 'overview'
            ? 'is-primary'
            : ''}"
          @click=${() => this.navigateTo('overview')}
        >
          Overview
        </button>
        <button
          class="button mt-5 ${this.tab === 'context' ? 'is-primary' : ''}"
          @click=${() => this.navigateTo('context')}
        >
          Context
        </button>
        <button
          class="button mt-5 ${this.tab === 'helpers' ? 'is-primary' : ''}"
          @click=${() => this.navigateTo('helpers')}
        >
          Helpers
        </button>
        <button
          class="button mt-5 ${this.tab === 'an-sso' ? 'is-primary' : ''}"
          @click=${() => this.navigateTo('an-sso')}
        >
          AN.sso
        </button>

        <slot class="box m-5"></slot>
      </omni-style>
    `;
  }
}

OmniElement.register('my-app', MyApp);
