// omni-app-container/demo/view-an-sso.js
import { html, OmniElement, OmniStyleElement } from 'omni-ui';

OmniStyleElement.register();

export class ViewAnSSO extends OmniElement {
  static properties = {
    showClients: { state: true },
    showProjects: { state: true },
  };

  render() {
    return html`
      <omni-style>
        <p>
          The portal (omni) app sets sso related data on the window object at
          <code>window.AN.sso</code>. This is globally available by JavaScript.
          The app container sets a similar object that is a subset of what
          portal sets. If something is missing that you need reach out to omni
          app container developers. Below shows what is set for you. This can
          make accessing certain sso related details very easy and works both in
          portal and app container with the same code.
        </p>

        <h3 class="mt-6 has-text-weight-bold is-size-2">AN.sso.session</h3>
        <pre>${JSON.stringify(window.AN.sso.session, null, 4)}</pre>

        <h3 class="mt-6 has-text-weight-bold is-size-2">AN.sso.user</h3>
        <pre>
${JSON.stringify(
            {
              ...window.AN.sso.user,
              clients: '(see below)',
              projects: '(see below)',
            },
            null,
            4
          )}</pre
        >

        <h3 class="mt-6 has-text-weight-bold is-size-2">AN.sso.user.clients</h3>
        <button
          class="button is-primary is-small m-5"
          @click=${() => {
            this.showClients = !this.showClients;
          }}
        >
          ${this.showClients ? 'Hide' : 'Show'}
        </button>
        <pre class="${this.showClients ? '' : 'is-hidden'}">
${JSON.stringify(window.AN.sso.user.clients, null, 4)}</pre
        >

        <h3 class="mt-6 has-text-weight-bold is-size-2">
          AN.sso.user.projects
        </h3>
        <button
          class="button is-primary is-small m-5"
          @click=${() => {
            this.showProjects = !this.showProjects;
          }}
        >
          ${this.showProjects ? 'Hide' : 'Show'}
        </button>
        <pre class="${this.showProjects ? '' : 'is-hidden'}">
${JSON.stringify(window.AN.sso.user.projects, null, 4)}</pre
        >
      </omni-style>
    `;
  }
}

OmniElement.register('view-an-sso', ViewAnSSO);
